### ISSUE TITLE
How the testevdev test cases are produced

### ISSUE THREAD

**Issue by iacore** [libsdl-org/SDL#7801]
I saw the commits below

- https://github.com/libsdl-org/SDL/commit/2fb1df155138accdf77700ee7f5a76d4bd0895ac

How do I add a new controller test case for this?

**Comment by slouken** [comment]
@smcv can help you here

**Comment by slouken** [comment]
Yes, the latest version of SDL will match USB VID/PID if the version doesn't match.

**Comment by iacore** [comment]
Closing for now. If anyone knows how to produce test data (see first comment) please tell.

**Comment by smcv** [comment]
That test data is currently only used to test `SDL_EVDEV_GuessDeviceClass()`, which is about the evdev capability bits, not the USB vendor/product IDs - so if your device is being mis-detected based on its vendor/product IDs, this is probably not going to help you yet. Our test data does include the vendor/product, but currently only as reference information, and they're not actually inputs to the function under test.

I'd like to expand our test coverage to make more use of this data to test other functions over time, though!

The test data mostly comes from the output of the `evemu-describe` tool. For example, output for the laptop keyboard that I'm using to type this comment is quoted below, and that happens to be the same one I used to provide `.name = "Thinkpad T520 and X280 keyboards"` in SDL's test data. Because it's a keyboard, and access to those is restricted to prevent keyloggers, I had to use `sudo evemu-describe` - but for gamepads you will generally have access to the evdev device node (and if you don't, SDL will be unable to use that gamepad), so you won't need `sudo`.

If you're interested in providing more test data for an unusual device, you can either translate it yourself and open a pull request, or post the `evemu-describe` output on an issue and I'll try to turn it into test data at some point. If you open a pull request, please leave the source data as a comment on the pull request, so that we have something to go back to if we're not sure whether your translation into the test data format was correct.

In general we don't need test data for ordinary Xbox/Playstation-style gamepads (particularly third-party clones of Microsoft/Sony/Nintendo controllers), unless they're something that SDL mis-detects. They can usually be grouped into very similar devices that only differ by their vendor/product/version, and if you look through the commit history you'll see that I added some of my colleague's test devices as comments rather than machine-readable data, because there's not much point in repeating the unit tests on a functionally equivalent device.

If you have a device that is unusual, like a specialized simulation controller, then test data for that is a lot more interesting, because our heuristics for classifying devices might give the wrong answer for those. When a console manufacturer introduces a new generation (for instance the Playstation 6 will presumably have a new controller type), that's often also interesting to have listed, since it often has features that the previous generation didn't.

```
$ sudo evemu-describe /dev/input/event0
# EVEMU 1.3
...
# Supported events:
#   Event type 0 (EV_SYN)
...
#   Event type 1 (EV_KEY)
...
#   Event type 4 (EV_MSC)
...
#   Event type 17 (EV_LED)
...
#   Event type 20 (EV_REP)
...
```

The "Supported events" become the `.ev` field. The mapping to a list of hex bytes is not instantly obvious (you have to calculate it based on the `EV_` codes from `/usr/include/linux/input-event-codes.h`), but in practice it's usually possible to copy/paste from a different device that supports the same list of event types. Most gamepads have `SYN/KEY/ABS/MSC` and sometimes `FF`.

```
I: 0011 0001 0001 ab54
```

This translates to `.bus_type: 0x0011, .vendor: 0x0001, .product: 0x0001, .version: 0xab54`.

```
P: 00 00 00 00 00 00 00 00
```

If it wasn't all zeroes, we'd put this in `.props` after reformatting it as a list of C hex integers. A real example is that the Dual Shock touchpad is `P: 05 00 00 00 00 00 00 00` which becomes `.props = { 0x05 }` (trailing zeroes don't need to be written out explicitly).

```
B: 00 0b 00 00 00 00 00 00 00
```

The line starting with `B: 00` is not useful, as far as I know: it probably only depends on your kernel version, and not the actual device.

```
B: 01 fe ff ff ff ff ff ff ff
B: 01 ff ff ef ff df ff ff fe
B: 01 01 d0 00 f8 78 30 80 03
B: 01 00 00 00 02 04 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
B: 01 00 00 00 00 00 00 00 00
```

All the lines starting with `B: 01` are button/key codes, and turn into the `.keys` field in the test data (remove the `B: 01` prefix, and reformat the rest into C hex integers). Any trailing zeroes can be omitted, and we have some macros for shorthand representations of large blocks of zeroes.

```
B: 02 00 00 00 00 00 00 00 00
```

Everything after `B: 02` is a relative axis (`EV_REL`), and would go in `.rel` after removing the `B: 02` prefix, if it was non-zero. This keyboard doesn't have any relative axes, but mice do.

```
B: 03 00 00 00 00 00 00 00 00
```

Similarly everything after `B: 03` is an absolute axis, `EV_ABS` (joysticks, touchpads, touchscreens)...

```
B: 04 10 00 00 00 00 00 00 00
```

... `B: 04` is `EV_MSC`, miscellaneous axes, which are not interesting to SDL so we don't transcribe them ...

```
B: 05 00 00 00 00 00 00 00 00
```

... `B: 05` is `EV_SW`, switches, which again are not interesting to SDL ...

```
B: 11 07 00 00 00 00 00 00 00
```

... `B: 11` is `EV_LED`, keyboard LEDs, which again are not interesting ...

```
B: 12 00 00 00 00 00 00 00 00
```

... `B: 12` is `EV_SND`, sound devices (sound card jack-detection and things like that), which again are not interesting ...

```
B: 14 03 00 00 00 00 00 00 00
```

... keyboard autorepeat ...

```
B: 15 00 00 00 00 00 00 00 00
B: 15 00 00 00 00 00 00 00 00
```

... and force feedback/haptics, which are certainly interesting for gamers, but not very interesting from the point of view of identifying whether something is a joystick or not, so we don't transcribe these either.

Steam's `steam-runtime-input-monitor` tool will give you *most* of this information, but in a slightly different format, and it can unfortunately be incomplete for some devices. One of the various things on my to-do list is to try to fix it to provide complete information, to make it more convenient for Steam users to provide test data like this.

**Comment by iacore** [comment]
Is it possible for you to enable Github discussion and turn this into a discussion thread?
